src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -178,13 +163,34 @@ public TextDocumentState UpdateText(SourceText newText, PreservationMode mode)
     public TextDocumentState UpdateText(TextLoader loader, PreservationMode mode)
     {
         // don't blow up on non-text documents.
-        var newTextSource = mode == PreservationMode.PreserveIdentity
-            ? CreateStrongText(loader)
-            : CreateRecoverableText(loader, solutionServices);
+        var newTextSource = CreateTextFromLoader(loader, mode, this.solutionServices);
 
         return UpdateText(newTextSource, mode, incremental: false);
     }
 
+    private static ITextAndVersionSource CreateTextFromLoader(TextLoader loader, PreservationMode mode, SolutionServices solutionServices)
+    {
+        var service = solutionServices.GetRequiredService<IWorkspaceConfigurationService>();
+        var options = service.Options;
+
+        // If the caller is explicitly stating that identity must be preserved, then we created a source that will load
+        // from the loader the first time, but then cache that result so that hte same result is *always* returned.
+        if (mode == PreservationMode.PreserveIdentity || options.DisableRecoverableText)
+            return new LoadableTextAndVersionSource(loader, cacheResult: true);
+
+        // If the loader asks us to always hold onto it strongly, then we do not want to create a recoverable text
+        // source here.  Instead, we'll go back to the loader each time to get the text.  This is useful for when the
+        // loader knows it can always reconstitute the snapshot exactly as it was before.  For example, if the loader
+        // points at the contents of a memory mapped file in another process.
+        if (loader.AlwaysHoldStrongly)
+            return new LoadableTextAndVersionSource(loader, cacheResult: false);
```

Comment:
the thrust of this change is that we're doing the same as the line below, just not wrapping the source in a RecoverableTextAndVersion.  We have no need from that as our loader itself can always 'recover' the contents with ease.  This means we no longer dump the in-memory files to memory-mapped files in the OOP process.